### PR TITLE
Allow BigFloat to be deserialized from a JSON string

### PR DESCRIPTION
**Pull request by Sija**
Followup to #9898 as per https://github.com/crystal-lang/crystal/pull/9898#discussion_r538354577

### CODE REVIEW

**Review by straight-shoota — CHANGES_REQUESTED** [review]
This should be enhanced by specs and proper handling for unexpected token types.

**Comment by Sija** [comment]
@straight-shoota I've added specs as requested. Unsupported types will just raise so IMO there's nuffin' moar to do there.

**Review by straight-shoota — APPROVED** [review]
Sounds good.

I double checked how this is implemented for other types and it seems they all just try to read the expected type.

**Comment by straight-shoota** [comment]
I've looked at existing implementations and yeah, they all work that way that they just raise. This is fine for 1-to-1 mappings between JSON and Crystal types, but when the deserialization accepts multiple JSON types, only the last one will be mentioned in the error message. So there's room for improvement.

**Review by straight-shoota — APPROVED** [review]

File: src/big/json.cr [hunk 1]
```diff
@@ -2,20 +2,6 @@ require "json"
...
-      @io << number
-    end
-  end
-
-  # Writes a big float.
-  def number(number : BigFloat)
-    scalar do
-      @io << number
-    end
-  end
```
Comment (Sija) [inline src/big/json.cr:17]: ftr, this erroneously overshadowed `JSON::Builder#number(number : Float)` which has a custom logic for raising on illegal float (like `NaN` and `Infinity`).

Comment (asterite) [reply src/big/json.cr:17]: But there's no NaN or Infinity in BigFloat... 🤔

Comment (Sija) [reply src/big/json.cr:17]: Hmm, atm that seems true, I'm not sure though if that's intended...

```cr
require "big"

Float64::INFINITY.to_big_f # => Program received and didn't handle signal FPE (8)
Float64::NAN.to_big_f      # => Program received and didn't handle signal FPE (8)
```

Comment (asterite) [reply src/big/json.cr:17]: Okay, that's a different bug. I just don't think you should remove this chunk of code.

Comment (straight-shoota) [reply src/big/json.cr:17]: Well the generic implementation isn't *wrong* either. BigFloat answers `nil?` and `infinite?` with `false`, so the special branches won't match.

Comment (asterite) [reply src/big/json.cr:17]: Sounds good

**Comment by Sija** [comment]
GTG?

**Comment by Sija** [comment]
ping 🏓

**Pull request merged**